routing/datasource.go

```diff
@@ -0,0 +1,197 @@
+package routing
+
+import (
+	"fmt"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/filters"
+	"log"
+	"net/url"
+	"time"
+)
+
+type incomingType uint
+
+const (
+	incomingReset incomingType = iota
+	incomingUpdate
+)
+
+type routeDefs map[string]*eskip.Route
+
+type incomingData struct {
+	typ            incomingType
+	client         DataClient
+	upsertedRoutes []*eskip.Route
+	deletedIds     []string
+}
+
+func receiveInitial(c DataClient, pollTimeout time.Duration, out chan<- *incomingData) {
+	for {
+		routes, err := c.GetInitial()
+		if err != nil {
+			log.Println("error while receiveing initial data;", err)
+			time.Sleep(pollTimeout)
+			continue
+		}
+
+		out <- &incomingData{incomingReset, c, routes, nil}
+		return
+	}
+}
+
+func receiveUpdates(c DataClient, pollTimeout time.Duration, out chan<- *incomingData) {
+	for {
+		time.Sleep(pollTimeout)
+		routes, deletedIds, err := c.GetUpdate()
+		if err != nil {
+			log.Println("error while receiving update;", err)
+			return
+		}
+
+		if len(routes) > 0 || len(deletedIds) > 0 {
+			out <- &incomingData{incomingUpdate, c, routes, deletedIds}
+		}
+	}
+}
+
+func receiveFromClient(c DataClient, pollTimeout time.Duration, out chan<- *incomingData) {
+	for {
+		receiveInitial(c, pollTimeout, out)
+		receiveUpdates(c, pollTimeout, out)
+	}
+}
+
+func applyIncoming(defs routeDefs, d *incomingData) routeDefs {
+	if d.typ == incomingReset || defs == nil {
+		defs = make(routeDefs)
+	}
+
+	if d.typ == incomingUpdate {
+		for _, id := range d.deletedIds {
+			delete(defs, id)
+		}
+	}
+
+	if d.typ == incomingReset || d.typ == incomingUpdate {
+		for _, def := range d.upsertedRoutes {
+			defs[def.Id] = def
+		}
+	}
+
+	return defs
+}
+
+func mergeDefs(defsByClient map[DataClient]routeDefs) []*eskip.Route {
+	mergeById := make(routeDefs)
+	for _, defs := range defsByClient {
+		for id, def := range defs {
+			mergeById[id] = def
```

Comment:
discussed to make it defined. But I'd be fine with making that change in a separate, later PR